### ISSUE TITLE
URGENT: *TypeScript compiler* build is failing with new update

### ISSUE THREAD

**Issue by nonara** [octokit/plugin-rest-endpoint-methods.js#172]
**What happened?** 
TypeScript compiler depends on this, and it appears that the latest patch has changed some type information which is causing TS compiler build process to fail for versions 3.9+

**Errors**
```
[21:44:13] Finished 'generateLibs' after 571 ms
node_modules/@octokit/plugin-rest-endpoint-methods/dist-types/generated/parameters-and-response-types.d.ts:961:60 - error TS2339: Property 'GET /user/:migration_id/repositories' does not exist on type 'Endpoints'.

961             parameters: RequestParameters & Omit<Endpoints["GET /user/:migration_id/repositories"]["parameters"], "baseUrl" | "headers" | "mediaType">;
                                                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@octokit/plugin-rest-endpoint-methods/dist-types/generated/parameters-and-response-types.d.ts:962:33 - error TS2339: Property 'GET /user/:migration_id/repositories' does not exist on type 'Endpoints'.

962             response: Endpoints["GET /user/:migration_id/repositories"]["response"];
                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@octokit/plugin-paginate-rest/dist-types/generated/paginating-endpoints.d.ts:987:31 - error TS2339: Property 'GET /user/:migration_id/repositories' does not exist on type 'Endpoints'.

987         parameters: Endpoints["GET /user/:migration_id/repositories"]["parameters"];
                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@octokit/plugin-paginate-rest/dist-types/generated/paginating-endpoints.d.ts:988:29 - error TS2339: Property 'GET /user/:migration_id/repositories' does not exist on type 'Endpoints'.

988         response: Endpoints["GET /user/:migration_id/repositories"]["response"];
                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 4 errors.
```

**Comment by gr2m** [comment]
Odd, can you make sure if you have the latest version of @octokit/types in your dependency tree?

**Comment by sparksp** [comment]
> Odd, can you make sure if you have the latest version of @octokit/types in your dependency tree?

For me, I can see three major versions of @octokit/types in the lockfile (2, 4 and 5). `@octokit/rest@^16.43.1` wants `@octokit/plugin-rest-endpoint-methods=2.4.0` which in turn wants `@octokit/types^2.0.1`. If I add @octokit/types to my dependencies then everything is happy again.

**Comment by lukehedger** [comment]
I am seeing this same error with the following versions:

```json
"devDependencies": {
  "@octokit/types": "5.1.1",
  "typescript": "3.9.7"
},
"dependencies": {
  "@octokit/rest": "18.0.1"
}
```

Error log:

```
$ tsc
../node_modules/@octokit/plugin-paginate-rest/dist-types/generated/paginating-endpoints.d.ts:987:31 - error TS2339: Property 'GET /user/:migration_id/repositories' does not exist on type 'Endpoints'.

987         parameters: Endpoints["GET /user/:migration_id/repositories"]["parameters"];
                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

../node_modules/@octokit/plugin-paginate-rest/dist-types/generated/paginating-endpoints.d.ts:988:29 - error TS2339: Property 'GET /user/:migration_id/repositories' does not exist on type 'Endpoints'.

988         response: Endpoints["GET /user/:migration_id/repositories"]["response"];
                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

../node_modules/@octokit/plugin-rest-endpoint-methods/dist-types/generated/parameters-and-response-types.d.ts:961:60 - error TS2339: Property 'GET /user/:migration_id/repositories' does not exist on type 'Endpoints'.

961             parameters: RequestParameters & Omit<Endpoints["GET /user/:migration_id/repositories"]["parameters"], "baseUrl" | "headers" | "mediaType">;
                                                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

../node_modules/@octokit/plugin-rest-endpoint-methods/dist-types/generated/parameters-and-response-types.d.ts:962:33 - error TS2339: Property 'GET /user/:migration_id/repositories' does not exist on type 'Endpoints'.

962             response: Endpoints["GET /user/:migration_id/repositories"]["response"];
                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Found 4 errors.
```

**Comment by nonara** [comment]
@gr2m To clarify, I'm not talking about `tsc` failing for a project that I've written which uses the plugin. I'm referring to the source for the TypeScript compiler, itself, which has broad impact.

All versions above 3.8 are failing to build: https://github.com/microsoft/TypeScript/issues/39712

These are old releases which haven't been updated in awhile. This seems to have happened as of the update for this plugin yesterday. 

Steps to reproduce can be found in the issue above.

**Comment by gr2m** [comment]
oh my, I'm looking into it

**Comment by gr2m** [comment]
I was able to reproduce the problem. I think I know what happened, please give me a moment to verify.

**Comment by gr2m** [comment]
The problem should be fixed now. You need to update to latest `@octokit/rest` (18.0.2) and also make sure that to have the latest `@octokit/plugin-paginate-rest` sub dependency (2.2.4)

```
➜  TypeScript git:(master) ✗ npm ls @octokit/plugin-paginate-rest
typescript@4.0.0 /private/var/folders/hs/x9qtfmvn1lz1sgml9q21h7k80000gn/T/tmp.5Ggh7uap/TypeScript
└─┬ @octokit/rest@18.0.2
  └── @octokit/plugin-paginate-rest@2.2.4 
```

Can you verify?

I'll write up a post-mortem to explain what happened, how I'll make sure it won't happen again.

**Comment by jamesrenaud** [comment]
I was having the identical issue to @nonara . Updated to `@octokit/rest@18.0.2` and builds are completing.

```
"devDependencies": {
  "@octokit/types": "5.1.1",
  "typescript": "3.9.7"
},
"dependencies": {
  "@octokit/rest": "18.0.2"
}
```

**Comment by nonara** [comment]
Thanks for the fast turnaround, @gr2m! Confirming now...

**Comment by nonara** [comment]
Fresh clone compiles! 🎉

**Comment by gr2m** [comment]
Thank you for confirming!

**Comment by orta** [comment]
thanks folks

**Comment by kaaloo** [comment]
Awesome work! :+1:

**Comment by gr2m** [comment]
# Post Mortem: "Property 'GET /user/:migration_id/repositories' does not exist on type 'Endpoints'" TypeScript error

- **Impact**: TypeScript compilation errors. No production errors.
- **Packages**: `@octokit/rest` v18.x
- First report: July 22, 7:16pm PDT
- Resolved: Jul 23, 11:01am PDT

## Summary

Starting July 22, 4:20pm PDT, TypeScript users of the latest `@octokit/rest` version (18.x) started to see the following errors when compiling their code:

> Property 'GET /user/:migration_id/repositories' does not exist on type 'Endpoints'

The error originated from two `@octokit/rest` dependencies.

1. `@octokit/plugin-rest-endpoint-methods`
1. `@octokit/plugin-paginate-rest`

The problem was originally reported by the TypeScript team on July 22, 7:16pm PDT ([issue](https://github.com/octokit/plugin-rest-endpoint-methods.js/issues/172)). The problem was confirmed to be resolved on Jul 23, 11:01am PDT ([comment](https://github.com/octokit/plugin-rest-endpoint-methods.js/issues/172#issuecomment-663150380))

## What happened

The source of the problem was a GitHub REST API path correction, which was published in the `@github/openapi` package version 5.0.2 on July 22, 12:02pm PDT.

- Before (5.0.1): https://unpkg.com/browse/@github/openapi@5.0.1/dist/api.github.com.json#L50802
- After (5.0.2): https://unpkg.com/browse/@github/openapi@5.0.2/dist/api.github.com.json#L49639

Sever `@octokit/*` libraries receive automated pull request whenever a new version of the `@github/openapi` package is released. One such pull request updated `@octokit/types` in [octokit/types.ts#137](https://github.com/octokit/types.ts/pull/137), which changed the `/user/{migration_id}/repositories` path to `/user/migrations/{migration_id}/repositories`. Once merged, the new version 5.1.1 was published to npm. We do not consider typescript changes breaking changes, as they do not impact code run in production, only at build time during local development or CI/CD.

Another pull request was for `@octokit/plugin-rest-endpoints` (octokit/plugin-rest-endpoint-methods.js#171)[https://github.com/octokit/plugin-rest-endpoint-methods.js/pull/171], in which the `@octokit/types` dependency was updated besides the path updates, in order to keep the two in sync.

The problem at this is that while `@octokit/rest` has a pinned dependency on `@octokit/plugin-rest-endpoints`, `octokit/plugin-rest-endpoints` does not have a pinned dependency on `@ocotkit/types`, in order to allow npm to de-duplicate it as it's required by many `@octokit/*` packages. That's who `@octokit/plugin-rest-endpoints` and `@ocotkit/types` became out of sync in the dependency tree of `@octokit/rest`.

The problem with the `@octokit/plugin-rest-endpoints` dependency was resolved via [octokit/rest.js#1811](https://github.com/octokit/rest.js#1811) and a new version was released: `@octokit/rest@18.0.2`.

On top of that, the automated update of `@octokit/plugin-paginate-rest` triggered by the new `@github/openapi` release triggered silently: https://github.com/octokit/plugin-paginate-rest.js/runs/899898279?check_suite_focus=true#step:6:14. `@octokit/plugin-paginate-rest` uses `@octokit/types` for internal type definitions, too. That is how `@octokit/plugin-paginate-rest` and `@octokit/types` became out of sync. The problem was resolved via [octokit/plugin-paginate-rest.js#110](https://github.com/octokit/plugin-paginate-rest.js/pull/110) and a new version was released: `@octokit/plugin-paginate-rest@2.2.4`.

With `@octokit/rest@18.0.2` and `@octokit/plugin-paginate-rest@2.2.4`, the "Property 'GET /user/:migration_id/repositories' does not exist on type 'Endpoints'" TypeScript error no longer occurs.

## How we will prevent the problem from occuring again

In future, we will consider the removal or change of an existing API in `@octokit/types` to be a breaking chaneg. While we won't consider TypeScript changes to be breaking changes overall, as they do not impact code run in production, this incident showed that due to dependency on `@ocotkit/types` across many `@octokit/*` packages, the risk of out-of-sync packages is to high.

Had we published `@octokit/types@5.1.1` as `@octokit/types@6.0.0` instead, none of the problems had occured. The wors impact would be that different versions might have exited in the `@ocotkit/rest` dependency tree for a short amount of time.

As we will be more strict with breaking changes on `@octokit/types` moving forward, we will also invest time to avoid breaking changes in the first place, by implementing deprecations, as we do in our JavaScript packages such as `@octokit/rest` and `@octokit/plugin-rest-endpoint-methods`. Breaking changes will only occur to remove previously deprecated APIs.

The other problem this incident surfaced was the silent error at https://github.com/octokit/plugin-paginate-rest.js/runs/899898279?check_suite_focus=true#step:6:14. We will make sure that the GitHub OpenAPI update builds will fail in future, which will trigger a notification.

---

Please comment below if you have any further questions.
